source-dynamodb/backfill.go

```diff
@@ -0,0 +1,322 @@
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/time/rate"
+)
+
+func (c *capture) backfill(ctx context.Context, t *table, dur time.Duration) error {
+	var scanLimit *int32
+	if t.scanLimitFromConfig != 0 {
+		// Stick with the scan limit set by the advanced configuration, if present.
+		scanLimit = aws.Int32(int32(t.scanLimitFromConfig))
+	} else if t.effectiveRcus > 100 || t.effectiveRcus == 0 {
+		// For tables with a significant amount of provisioned capacity (or unlimited) do not
+		// restrict the number of returned items. There is still a 1 MB limit to the amount of data
+		// that a single scan request can return.
+		scanLimit = nil
+	} else {
+		// Tables with low provisioned RCUs use a lower limit to even out the load on the table.
+		scanLimit = aws.Int32(lowRcuScanLimit)
+	}
+
+	// Try to match the read capacity units consumed with either the configured RCU limit or the
+	// provisioned RCUs for the table.
+	rcuLimiter := rate.NewLimiter(rate.Limit(t.effectiveRcus), 1)
+	if t.effectiveRcus == 0 {
+		// If no RCU limit was set or determined from the table there will be no limit.
+		rcuLimiter.SetLimit(rate.Inf)
+	}
+
+	// Initialize the active segments list. We cycle through reads of segments using a FIFO queue to
+	// more evenly distribute reads across table partitions, as opposed to something like repeatedly
+	// reading the same segment over and over. Individual table partitions have a hard RCU cap so it
+	// is best to distribute read load across the table's partitions.
+	t.activeSegments = make(chan int, t.totalBackfillSegments)
+	for segment := 0; segment < t.totalBackfillSegments; segment++ {
+		if c.getSegmentState(t.tableName, segment).FinishedAt.IsZero() {
+			t.activeSegments <- segment
+		}
+	}
+
+	// Arrange for stopCh to be closed after the provided duration to signal the backfill workers to
+	// stop.
+	stopCh := make(chan struct{})
+	stopTimer := time.AfterFunc(dur, func() { close(stopCh) })
+	defer stopTimer.Stop()
+
+	// Run the backfill workers. They will continuously pull segements from t.activeSegements and
+	// run backfill scans, returning only if they run out of segments or stopCh is closed.
+	eg, groupCtx := errgroup.WithContext(ctx)
+	for idx := 0; idx < backfillConcurrency; idx++ {
```

Comment:
we'll spawn three backfill workers per-binding, no matter how many bindings there are?

If I'm back-filling from 10 tables, that's 30 concurrent backfill requests, which I would expect to fight with each other for CPU. It seems pretty easy to get into situations where this would perform poorly, as compared to fixed parallelism of backfills across _all_ tables, no?

Or is the implicit assumption that there just aren't many bindings, given how discovery works (haven't gotten there yet)?

Reply:
The main thing I grappled with for this is if there was a table with a low provisioned capacity, the workers for that table might spend a lot of time just sitting around waiting for the rate limiter to refill. So with a fixed worker pool the entire pool could end up filled with workers doing nothing but waiting intermittently.

But you're right If there are a lot of tables, this wouldn't be optimal. I'll see if I can come up with another way to structure this. One thing that comes to mind is using an additional semaphore that a worker must acquire after awaiting the rate limiter for the table, but prior to actually doing the work.

Reply:
ah! okay, that's important context. I'm happy to merge as-is and see where the actual problems shake out.